tools/python/mwm_downloader.py

```diff
@@ -0,0 +1,152 @@
+#!/usr/bin/env python3
+
+import argparse
+import os
+import re
+import time
+import sys
+import urllib.request
+import urllib.parse
+import urllib.error
+import logging
+
+from concurrent.futures import ThreadPoolExecutor, as_completed
+
+
+DIRECT_MAP_URL = 'http://direct.mapswithme.com/direct/'
+ROOT = os.path.dirname(os.path.realpath(__file__))
+DEFAULT_DOWNLOAD_DIRECTORY = os.path.join(os.path.dirname(os.path.dirname(ROOT)), 'data')
+MAP_REVISION_REGEXP = re.compile(r'(\d{6})')
+MWM_NAME_REGEXP = re.compile(r'"(\S+\.mwm)"')
+
+logger = logging.getLogger(__name__)
+
+
+def parse_html_by_regexp(url, regexp):
+    def parse_line(line):
+        return regexp.search(line.decode('utf-8'))
+
+    with urllib.request.urlopen(url) as f:
+        lines = f.readlines()
+        filtered_lines = filter(None, map(parse_line, lines))
+        return (line.group(1) for line in filtered_lines)
+
+
+def get_mwm_versions():
+    return map(int, parse_html_by_regexp(DIRECT_MAP_URL, MAP_REVISION_REGEXP))
+
+
+def get_mwm_names(version_num):
+    version_url = '{base}{version}'.format(base=DIRECT_MAP_URL, version=version_num)
+    return map(urllib.parse.unquote, parse_html_by_regexp(version_url, MWM_NAME_REGEXP))
+
+
+def download_file(url, filename, attempts=3):
+    try:
+        urllib.request.urlretrieve(url, filename=filename)
+    except (urllib.error.URLError, PermissionError):
```

Comment:
The problem here is that different types of errors mean different handlers. For example, if you recieve a 4xx error you probably don't want a retry at all.

Reply:
Fixed.